### DISCUSSION TITLE
How can I setup custom useragent for each isolated browser instance?

### DISCUSSION THREAD

**Discussion by assopri** [cefsharp/CefSharp#3577] [Q&A]
I do not use Cef initialize, so I cannot install useragent using Cefsettings. To init browser (Winforms) I use the following code:
`  BrowserSettings browserSetting new BrowserSettings ();
            Browser = new ChromiumWebBrowser();
            (Browser as ChromiumWebBrowser).BrowserSettings = browserSetting;
            RequestContextSettings requestContextSettings = new RequestContextSettings();
            requestContextSettings.PersistSessionCookies = true;
            requestContextSettings.PersistUserPreferences = true;
            requestContextSettings.CachePath = @"C:\Users\User\Documents\temp";
            
            (Browser as ChromiumWebBrowser).RequestContext = new RequestContext(requestContextSettings);`

How can I make it use custom user agent when loading?

**Comment by amaitland** [comment]
The general usage guide covers a great deal of topics including setting user agent see https://github.com/cefsharp/CefSharp/wiki/General-Usage#useragent

I'd suggest have a browse through the https://github.com/cefsharp/CefSharp/wiki/General-Usage when you have time. 

---

Supporting and developing CefSharp takes a great deal of time and effort, if you are using CefSharp in a commercial application please consider funding my efforts. Even $5 a month makes a difference (Paying for twelve months in advance if you wish to make a single payment).

https://github.com/sponsors/amaitland/

**Reply by assopri** [reply]
Sure! Yesterday subscribed for 10$. You are doing great job!

**Reply by amaitland** [reply]
Thank you kindly, greatly appreciated 👍

**Reply by assopri** [reply]
> The general usage guide covers a great deal of topics including setting user agent see https://github.com/cefsharp/CefSharp/wiki/General-Usage#useragent

I use the 3rd approach of user agent overriding (with OnBeforeResourceLoad)
![image](https://user-images.githubusercontent.com/73235464/119805155-a92eb500-bee9-11eb-851e-55dd16a1e313.png) 

However, as far as I understand website can elicit my primordial useragent checking it in Javascript. Any workaround to handle this (so that if website still checks my user agent with JS, it would have got the same one I set in OnBeforeResourceLoad ) ?

**Reply by amaitland** [reply]
> However, as far as I understand website can elicit my primordial useragent checking it in Javascript. Any workaround to handle this (so that if website still checks my user agent with JS, it would have got the same one I set in OnBeforeResourceLoad ) ?

That is correct, setting the user agent header in OnBeforeResourceLoad only changes the header, in JavaScript the default user agent will still be used.

Currently using the DevTools approach is the only option for changing the user agent at runtime.

**Reply by assopri** [reply]
If I use this code - will Javascript still be able to identify my default user agent?

```
using (var client = chromiumWebBrowser.GetDevToolsClient())
{
	_ = client.Network.SetUserAgentOverrideAsync("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/87.0.4280.88 Safari/537.36 - Testing 123");
}
```

I also get Exception using it:

The ChromiumWebBrowser instance creates the underlying Chromium Embedded Framework (CEF) browser instance in an async fashion. The undelying CefBrowser instance is not yet initialized. Use the IsBrowserInitializedChanged event and check the IsBrowserInitialized property to determine when the browser has been initialized.

` Browser = new ChromiumWebBrowser();
           
            (Browser as ChromiumWebBrowser).BrowserSettings = browserSetting;
            //Using RequestContextSettings allows having isolated cache/cookies etc for different ChromiumWebBrowser instances
            RequestContextSettings requestContextSettings = new RequestContextSettings();
            requestContextSettings.PersistSessionCookies = true;
            requestContextSettings.PersistUserPreferences = true;
            
            string cachePath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.MyDocuments),
                UniCefBrowserWrapperFactory.CefSharpCacheLocalPath,
                SingleBrowserInfo.Proxy.Ip);

            if (!Directory.Exists(cachePath)) Directory.CreateDirectory(cachePath);
            requestContextSettings.CachePath = cachePath;



            (Browser as ChromiumWebBrowser).RequestContext = new RequestContext(requestContextSettings);

            (Browser as ChromiumWebBrowser).RequestHandler =
                new CustomRequestHandler(SingleBrowserInfo.Proxy.Login,
                    SingleBrowserInfo.Proxy.Password,
                    SingleBrowserInfo.UserAgent, PartialLoad);
            
            using (var client = (Browser as ChromiumWebBrowser).GetDevToolsClient())
            {
                _ = client.Network.SetUserAgentOverrideAsync(" Testing 123");
            }`

**Reply by amaitland** [reply]
> If I use this code - will Javascript still be able to identify my default user agent?

The user agent for the browser will be changed, this includes the user agent reported to JavaScript. 

> I also get Exception using it:

As the exception says you are calling the method too soon. You can use the http://cefsharp.github.io/api/91.1.x/html/E_CefSharp_WinForms_ChromiumWebBrowser_IsBrowserInitializedChanged.htm event to set the user agent. The underlying browser will have been created, before the first page load.

https://github.com/cefsharp/CefSharp.MinimalExample/blob/master/CefSharp.MinimalExample.WinForms/BrowserForm.cs#L26

**Reply by assopri** [reply]
Cool! For screen version it works just fine. What is the analog for Offscreen?
![image](https://user-images.githubusercontent.com/73235464/125890707-19a09ff7-0183-4649-9f2e-727d4ecd9f1f.png)

**Reply by amaitland** [reply]
Using a delegate that's passed into the constructor of `CefSharp.OffScreen.ChromiumWebBrowser`.

https://github.com/cefsharp/CefSharp/issues/3552#issuecomment-846344387

**Reply by assopri** [reply]
Cool!
